R/map.R

```diff
@@ -218,6 +221,27 @@ map_raw <- function(.x, .f, ...) {
   .Call(map_impl, environment(), ".x", ".f", "raw")
 }
 
+
+#' @rdname map
+#' @param .ptype If `NULL`, the default, the output type is the common type
+#'   of the elements of the result. Otherwise, supply a "prototype" giving
+#'   the desired type of output.
+#' @importFrom vctrs vec_c vec_size vec_ptype_common
+#' @export
+map_vec <- function(.x, .f, ..., .ptype = NULL) {
+  out <- map(.x, .f, ...)
+
+  .ptype <- vec_ptype_common(!!!out, .ptype = .ptype)
+  for (i in seq_along(out)) {
+    if (vec_size(out[[i]]) != 1L) {
+      stop_bad_element_vector(out[[i]], i, .ptype, 1L, what = "Result")
+    }
+  }
+
+  vec_c(!!!out, .ptype = .ptype)
```

Comment:
Minor: I'd probably use `vec_unchop()` here.

Reply:
Should we pass a `name_spec` argument?

Reply:
I didn't do use `vec_unchop()` because the docs say "When no indices are supplied, it is identical to vec_c()." Is that not correct, or do you think think `vec_unchop()` conveys the semantics better?

Reply:
Avoiding the `!!!` typically is a little more memory efficient, not sure if there is an rlang reason why this is the case or not

``` r
library(vctrs)

x <- as.list(1:10000)

bench::mark(vec_unchop(x), vec_c(!!!x), iterations = 1000)[c("expression", "median", "mem_alloc")]
#> # A tibble: 2 × 3
#>   expression      median mem_alloc
#>   <bch:expr>    <bch:tm> <bch:byt>
#> 1 vec_unchop(x)   2.37ms    81.8KB
#> 2 vec_c(!!!x)     2.45ms   159.7KB
```

<sup>Created on 2022-08-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>